tcp/tcp.c

```diff
@@ -0,0 +1,382 @@
+		tcpNum = 0;
+	}
+
+	// allocate a connection structure
+	TcpClient *client = tcpClient + tcpNum;
+	uint8_t clientNum = tcpNum;
+	tcpNum = (tcpNum+1)%MAX_TCP;
+
+	// free any data structure that may be left from a previous connection
+	if (client->data) os_free(client->data);
+	if (client->pCon) {
+		if (client->pCon->proto.tcp) os_free(client->pCon->proto.tcp);
+		os_free(client->pCon);
+	}
+	os_memset(client, 0, sizeof(TcpClient));
+	DBG_TCP("TCP #%d: Setup host=%s port=%d \n", clientNum, tcp_host, port);
+
+	client->sock_mode = sock_mode;
+	client->resp_cb = cmd->value;
+	client->conn_num = clientNum;
+
+	client->host = (char *)tcp_host;
+	client->port = port;
+
+	client->pCon = (struct espconn *)os_zalloc(sizeof(struct espconn));
+	if (client->pCon == NULL) {
+		DBG_TCP("TCP #%d: Setup failed to alloc memory for client_pCon\n", clientNum);
+		goto fail;
+	}
+
+	client->pCon->type = ESPCONN_TCP;
+	client->pCon->state = ESPCONN_NONE;
+	client->pCon->proto.tcp = (esp_tcp *)os_zalloc(sizeof(esp_tcp));
+	if (client->pCon->proto.tcp == NULL) {
+		DBG_TCP("TCP #%d: Setup failed to alloc memory for client->pCon->proto.tcp\n", clientNum);
+		goto fail;
+	}
+
+	os_memcpy(client->host, tcp_host, 4);
+	client->pCon->proto.tcp->remote_port = client->port;
+	client->pCon->proto.tcp->local_port = client->port; // espconn_port();
+
+	client->pCon->reverse = client;
+
+	espconn_regist_sentcb(client->pCon, tcpclient_sent_cb);
+	espconn_regist_recvcb(client->pCon, tcpclient_recv_cb);
+	espconn_regist_reconcb(client->pCon, tcpclient_recon_cb);
+	if (client->sock_mode == SOCKET_SERVER) { // Server mode?
+		DBG_TCP("TCP #%d: Enable server mode on port%d\n", clientNum, client->port);
+		espconn_accept(client->pCon);
+		espconn_regist_connectcb(client->pCon, tcpclient_connect_cb);
+	}
+	
+	cmdResponseStart(CMD_RESP_V, clientNum, 0);
+	cmdResponseEnd();
+	DBG_TCP("TCP #%d: setup finished\n", clientNum);
+	return;
+
+fail:
+	cmdResponseStart(CMD_RESP_V, err, 0);
+	cmdResponseEnd();
+	return;
+}
+
+void ICACHE_FLASH_ATTR
+TCP_Send(CmdPacket *cmd) {
+	CmdRequest req;
+	cmdRequest(&req, cmd);
+	
+	// Get client
+	uint32_t clientNum = cmd->value;
+	TcpClient *client = tcpClient + (clientNum % MAX_TCP);
+	DBG_TCP("TCP #%d: send", clientNum);
+
+	if (cmd->argc != 1 && cmd->argc != 2) {
+		DBG_TCP("\nTCP #%d: send - wrong number of arguments\n", clientNum);
+		return;
+	}
+	
+	// Get data to sent
+	uint16_t dataLen = cmdArgLen(&req);
+	DBG_TCP(" dataLen=%d", dataLen);
+	char tcpData[1024];
+	cmdPopArg(&req, tcpData, dataLen);
+	tcpData[dataLen] = 0;
+	DBG_TCP(" tcpData=%s", tcpData);
+
+	// we need to allocate memory for the data. We copy the message into it
+	char *tcpDataSet = "%s";
+	
+	if (client->data) os_free(client->data);
+	client->data = (char*)os_zalloc(dataLen);
+	if (client->data == NULL) {
+		DBG_TCP("\nTCP #%d failed to alloc memory for client->data\n", clientNum);
+		goto fail;
+	}
+	client->data_len = os_sprintf((char*)client->data, tcpDataSet, tcpData);
```

Comment:
why are you using this? the cmdPopArg should receive the data straight into a malloc'ed buffer and there should be no intermediate copy, certainly no string copy, which prevents sending data with zeroes in it.